File: src/args.ts
export type CliOptions = {
  all: boolean;
  maxDepth: number;
  fullscreen: boolean;
  verbose: boolean;
  noFetch: boolean;
  dir: string;
  help: boolean;
  version: boolean;
};

export type ParseResult =
  | { ok: true; options: CliOptions }
  | { ok: false; error: string };

export const DEFAULT_MAX_DEPTH = 3;

export const USAGE = `Usage: gitsy [options]

Show git status across child repositories and linked worktrees.

Options:
  --all              Show all discovered repositories, including clean repos
  --max-depth <n>    Scan repository directories up to n nested levels (default: 3)
  --dir <path>       Start scanning from <path> instead of the current directory
  --fullscreen       Open a scrollable fullscreen terminal UI
  --verbose          Print warnings for skipped repos and failed git commands
  --no-fetch         Skip fetching upstream changes (use local status only)
  --help             Show this help message
  --version          Show package version
`;

export function parseArgs(argv: readonly string[]): ParseResult {
  const options: CliOptions = {
    all: false,
    maxDepth: DEFAULT_MAX_DEPTH,
    fullscreen: false,
    verbose: false,
    noFetch: false,
    dir: process.cwd(),
    help: false,
    version: false,
  };

  for (let index = 0; index < argv.length; index += 1) {
    const arg = argv[index];

    if (arg === "--") {
      continue;
    }

    if (arg === "--all") {
      options.all = true;
      continue;
    }

    if (arg === "--fullscreen") {
      options.fullscreen = true;
      continue;
    }

    if (arg === "--verbose") {
      options.verbose = true;
      continue;
    }

    if (arg === "--no-fetch") {
      options.noFetch = true;
      continue;
    }

    if (arg === "--dir") {
      const value = argv[index + 1];
      if (value === undefined || value.startsWith("--")) {
        return { ok: false, error: "Missing value for --dir" };
      }
      options.dir = value;
      index += 1;
      continue;
    }

    if (arg?.startsWith("--dir=")) {
      const value = arg.slice("--dir=".length);
      if (value === "") {
        return { ok: false, error: "Missing value for --dir" };
      }
      options.dir = value;
      continue;
    }

    if (arg === "--help") {
      options.help = true;
      continue;
    }

    if (arg === "--version") {
      options.version = true;
      continue;
    }

    if (arg === "--max-depth") {
      const value = argv[index + 1];
      if (value === undefined || value.startsWith("--")) {
        return { ok: false, error: "Missing value for --max-depth" };
      }
      const parsed = parseMaxDepth(value);
      if (parsed === undefined) {
        return { ok: false, error: `Invalid --max-depth value: ${value}` };
      }
      options.maxDepth = parsed;
      index += 1;
      continue;
    }

    if (arg?.startsWith("--max-depth=")) {
      const value = arg.slice("--max-depth=".length);
      const parsed = parseMaxDepth(value);
      if (parsed === undefined) {
        return { ok: false, error: `Invalid --max-depth value: ${value}` };
      }
      options.maxDepth = parsed;
      continue;
    }

    return { ok: false, error: `Unknown argument: ${arg}` };
  }

  return { ok: true, options };
}

function parseMaxDepth(value: string): number | undefined {
  if (!/^\d+$/.test(value)) {
    return undefined;
  }

  const parsed = Number.parseInt(value, 10);
  if (!Number.isSafeInteger(parsed) || parsed < 1) {
    return undefined;
  }

  return parsed;
}

File: internal/args/args.go
package args

import (
	"fmt"
	"strconv"
	"strings"
)

const DefaultMaxDepth = 3

const Usage = `Usage: gitsy [options]

Show git status across child repositories and linked worktrees.

Options:
  --all              Show all discovered repositories, including clean repos
  --max-depth <n>    Scan repository directories up to n nested levels (default: 3)
  --dir <path>       Start scanning from <path> instead of the current directory
  --verbose          Print warnings for skipped repos and failed git commands
  --no-fetch         Skip fetching upstream changes (use local status only)
  --sync             Fast-forward repos that can safely update without conflicts (always fetches)
  --help             Show this help message
  --version          Show package version
`

type Options struct {
	All      bool
	MaxDepth int
	Verbose  bool
	NoFetch  bool
	Sync     bool
	Dir      string
	Help     bool
	Version  bool
}

type ParseResult struct {
	OK      bool
	Options Options
	Err     error
}

func Parse(argv []string, cwd string) ParseResult {
	options := Options{
		MaxDepth: DefaultMaxDepth,
		Dir:      cwd,
	}

	for index := 0; index < len(argv); index++ {
		arg := argv[index]

		switch {
		case arg == "--":
			continue
		case arg == "--all":
			options.All = true
		case arg == "--verbose":
			options.Verbose = true
		case arg == "--no-fetch":
			options.NoFetch = true
		case arg == "--sync":
			options.Sync = true
		case arg == "--help":
			options.Help = true
		case arg == "--version":
			options.Version = true
		case arg == "--dir":
			value, ok := nextValue(argv, index)
			if !ok {
				return parseError("Missing value for --dir")
			}
			options.Dir = value
			index++
		case strings.HasPrefix(arg, "--dir="):
			value := strings.TrimPrefix(arg, "--dir=")
			if value == "" {
				return parseError("Missing value for --dir")
			}
			options.Dir = value
		case arg == "--max-depth":
			value, ok := nextValue(argv, index)
			if !ok {
				return parseError("Missing value for --max-depth")
			}
			parsed, ok := parseMaxDepth(value)
			if !ok {
				return parseError(fmt.Sprintf("Invalid --max-depth value: %s", value))
			}
			options.MaxDepth = parsed
			index++
		case strings.HasPrefix(arg, "--max-depth="):
			value := strings.TrimPrefix(arg, "--max-depth=")
			parsed, ok := parseMaxDepth(value)
			if !ok {
				return parseError(fmt.Sprintf("Invalid --max-depth value: %s", value))
			}
			options.MaxDepth = parsed
		default:
			return parseError(fmt.Sprintf("Unknown argument: %s", arg))
		}
	}

	return ParseResult{OK: true, Options: options}
}

func nextValue(argv []string, index int) (string, bool) {
	if index+1 >= len(argv) || strings.HasPrefix(argv[index+1], "--") {
		return "", false
	}
	return argv[index+1], true
}

func parseMaxDepth(value string) (int, bool) {
	if value == "" {
		return 0, false
	}
	for _, char := range value {
		if char < '0' || char > '9' {
			return 0, false
		}
	}
	parsed, err := strconv.Atoi(value)
	if err != nil || parsed < 1 {
		return 0, false
	}
	return parsed, true
}

func parseError(message string) ParseResult {
	return ParseResult{OK: false, Err: fmt.Errorf("%s", message)}
}
